File: apps/sdk-typescript/src/crypto/did-key.ts
/**
 * `did:key` Ed25519 helpers (W3C did:key spec).
 *
 * Mirror of `apps/agent-identity` + `packages/core-crypto/src/did-key.ts`,
 * inlined here so the SDK ships standalone.
 *
 * Format: `did:key:z<multibase-base58btc(0xed01 || pubkey32)>`
 *
 * `0xed 0x01` is the varint multicodec prefix for Ed25519 public keys.
 */

import * as ed from '@noble/ed25519';

import { fromBase58btc, toBase58btc, toBase64 } from './codec.js';
import { wireSha512 } from './sha512.js';

// `@noble/ed25519` v2 is hash-pluggable but ships without a default sha512.
// We wire it once at module load to a Node-native implementation. Wiring is
// idempotent — a second call from another module simply re-assigns the same
// function.
ed.etc.sha512Sync = wireSha512;
ed.etc.sha512Async = (...messages: Uint8Array[]) => Promise.resolve(wireSha512(...messages));

const ED25519_MULTICODEC_PREFIX = new Uint8Array([0xed, 0x01]);
const DID_KEY_PREFIX = 'did:key:';

export interface GeneratedDidKey {
  /** `did:key:z6Mk...` identifier ready to register with `agent-identity`. */
  readonly did: string;
  /** Raw 32-byte Ed25519 public key, base64-encoded (RFC 4648, with padding). */
  readonly publicKeyBase64: string;
  /** Raw 32-byte Ed25519 secret key, base64-encoded. KEEP SECRET. */
  readonly secretKeyBase64: string;
}

/**
 * Generate a fresh Ed25519 keypair and encode the public key as a `did:key`.
 * Returns the DID + base64-encoded public/secret keys ready to feed into
 * `client.identity.register({ publicKey: publicKeyBase64, ... })`.
 */
export const generateDidKey = async (): Promise<GeneratedDidKey> => {
  const secretKey = ed.utils.randomPrivateKey();
  const publicKey = await ed.getPublicKeyAsync(secretKey);
  return {
    did: encodeDidKey(publicKey),
    publicKeyBase64: toBase64(publicKey),
    secretKeyBase64: toBase64(secretKey),
  };
};

/** Encode a 32-byte Ed25519 public key into its `did:key` form. */
export const encodeDidKey = (publicKey: Uint8Array): string => {
  if (publicKey.length !== 32) {
    throw new Error(`Ed25519 public key must be 32 bytes, got ${publicKey.length}`);
  }
  const prefixed = new Uint8Array(ED25519_MULTICODEC_PREFIX.length + publicKey.length);
  prefixed.set(ED25519_MULTICODEC_PREFIX, 0);
  prefixed.set(publicKey, ED25519_MULTICODEC_PREFIX.length);
  return `${DID_KEY_PREFIX}z${toBase58btc(prefixed)}`;
};

export interface ParsedDidKey {
  readonly publicKey: Uint8Array;
  readonly publicKeyBase64: string;
}

/**
 * Decode a `did:key` identifier into its raw + base64 public-key forms.
 * Throws on malformed prefix, non-`z` multibase, or unknown multicodec.
 */
export const parseDidKey = (did: string): ParsedDidKey => {
  if (!did.startsWith(DID_KEY_PREFIX)) {
    throw new Error(`Not a did:key identifier: ${did}`);
  }
  const multibase = did.slice(DID_KEY_PREFIX.length);
  if (!multibase.startsWith('z')) {
    throw new Error(`Unsupported multibase encoding (expected 'z'): ${multibase[0] ?? ''}`);
  }
  const decoded = fromBase58btc(multibase.slice(1));
  if (decoded.length < 3) {
    throw new Error('Decoded did:key payload too short');
  }
  if (decoded[0] !== ED25519_MULTICODEC_PREFIX[0] || decoded[1] !== ED25519_MULTICODEC_PREFIX[1]) {
    throw new Error(
      `Unknown did:key multicodec prefix: 0x${(decoded[0] ?? 0).toString(16)}${(decoded[1] ?? 0).toString(16)}`,
    );
  }
  const publicKey = decoded.slice(ED25519_MULTICODEC_PREFIX.length);
  return { publicKey, publicKeyBase64: toBase64(publicKey) };
};

// Re-export for tests + consumer convenience.
export { fromBase64, toBase64 } from './codec.js';

File: apps/sdk-python/src/colber_sdk/crypto/did_key.py
"""``did:key`` Ed25519 helpers (W3C did:key spec).

Mirror of ``apps/sdk-typescript/src/crypto/did-key.ts`` and
``packages/core-crypto/src/did-key.ts``.

Format: ``did:key:z<multibase-base58btc(0xed01 || pubkey32)>``

``0xed 0x01`` is the varint multicodec prefix for Ed25519 public keys.
The Python implementation produces byte-identical DID strings to the TS
SDK on the same input keypair.
"""

from __future__ import annotations

from dataclasses import dataclass

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

from ._codec import from_base58btc, to_base58btc, to_base64

ED25519_MULTICODEC_PREFIX = bytes([0xED, 0x01])
DID_KEY_PREFIX = "did:key:"


@dataclass(frozen=True, slots=True)
class GeneratedDidKey:
    """A freshly minted DID + Ed25519 keypair.

    Attributes:
        did: ``did:key:z6Mk...`` identifier ready to register with the
            ``agent-identity`` service.
        public_key_b64: Raw 32-byte Ed25519 public key, base64-encoded
            (RFC 4648, with padding).
        secret_key_b64: Raw 32-byte Ed25519 secret key, base64-encoded.
            **KEEP SECRET.**
    """

    did: str
    public_key_b64: str
    secret_key_b64: str


@dataclass(frozen=True, slots=True)
class ParsedDidKey:
    """The decoded contents of a ``did:key`` identifier."""

    public_key: bytes
    public_key_b64: str


def generate_did_key() -> GeneratedDidKey:
    """Generate a fresh Ed25519 keypair and encode the public key as a ``did:key``.

    Returns the DID + base64-encoded public/secret keys ready to feed into
    ``client.identity.register(public_key=..., owner_operator_id=...)``.
    """
    private_key = Ed25519PrivateKey.generate()
    secret_bytes = private_key.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    )
    public_bytes = private_key.public_key().public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )
    return GeneratedDidKey(
        did=encode_did_key(public_bytes),
        public_key_b64=to_base64(public_bytes),
        secret_key_b64=to_base64(secret_bytes),
    )


def encode_did_key(public_key: bytes) -> str:
    """Encode a 32-byte Ed25519 public key into its ``did:key`` form.

    Raises:
        ValueError: when ``public_key`` is not exactly 32 bytes.
    """
    if len(public_key) != 32:
        raise ValueError(f"Ed25519 public key must be 32 bytes, got {len(public_key)}")
    prefixed = ED25519_MULTICODEC_PREFIX + public_key
    return f"{DID_KEY_PREFIX}z{to_base58btc(prefixed)}"


def parse_did_key(did: str) -> ParsedDidKey:
    """Decode a ``did:key`` identifier into its raw + base64 public-key forms.

    Raises:
        ValueError: on malformed prefix, non-``z`` multibase, truncated
            payload, or unknown multicodec prefix.
    """
    if not did.startswith(DID_KEY_PREFIX):
        raise ValueError(f"Not a did:key identifier: {did}")
    multibase = did[len(DID_KEY_PREFIX) :]
    if not multibase.startswith("z"):
        first = multibase[0] if multibase else ""
        raise ValueError(f"Unsupported multibase encoding (expected 'z'): {first}")
    decoded = from_base58btc(multibase[1:])
    if len(decoded) < 3:
        raise ValueError("Decoded did:key payload too short")
    if decoded[0] != ED25519_MULTICODEC_PREFIX[0] or decoded[1] != ED25519_MULTICODEC_PREFIX[1]:
        raise ValueError(f"Unknown did:key multicodec prefix: 0x{decoded[0]:02x}{decoded[1]:02x}")
    public_key = bytes(decoded[len(ED25519_MULTICODEC_PREFIX) :])
    return ParsedDidKey(public_key=public_key, public_key_b64=to_base64(public_key))
